### PR TITLE
chore: update solang-parser

### PR DESCRIPTION
**Pull request by DaniPopes**
## Motivation

Relevant changes:
- hyperledger/solang#1230
  - Fixes a panic that we had to for check ourselves in Chisel (dc6eec5c50fa1d5c309eab97282e759fea5ef55c)
- hyperledger/solang#1237
  - Upstreams many of the parser "ext" traits in `forge-fmt` and `forge-doc` (2017824efe9bf9ab3c28cbba765f58d9a097c01b)

~Waiting on solang-parser v0.2.4 release, draft until then.~

Blocked by ~gakonst/ethers-rs#2334 and~ an Ethers release.

### CODE REVIEW

**Review by mattsse — APPROVED** [review]
awesome!
love that we could simplify this

lgtm, pending release

**Comment by DaniPopes** [comment]
Looks like 2 different errors in tests:
1. `IfStatement/original.sol`: looks like a previous bug, where the if statement on line 89 wasn't getting inlined?
2. `InlineDisable/original.sol`: brackets get removed on empty blocks when disabled inline, I cannot figure out where this happens

cc @rkrasiuk @jpopesculian pls advise

**Comment by mattsse** [comment]
@rkrasiuk please take a look at the failing tests, this is now a blocker for any PR that needs new ethers release

**Comment by rkrasiuk** [comment]
@DaniPopes thanks for the update, you are an OG! looking at failing tests rn

**Comment by rkrasiuk** [comment]
regarding 2, i narrowed it down to the line below 
https://github.com/foundry-rs/foundry/blob/37d4625eb7a851a5076fc31636475753ed997574/fmt/src/formatter.rs#L2332
not sure yet under which circumstances the location is invalid, tried changing it to
```rust
return_source_if_disabled!(self, func.loc().with_end(body.loc().end()));
```
but that results in different errors 🤷‍♂️

**Comment by mattsse** [comment]
I've added a smaller repro `IfStatement2` and tracked it down to:

1. https://github.com/foundry-rs/foundry/blob/2e25e08d64169df2dc541b4d0a15b5790d068a96/fmt/src/formatter.rs#L1188 called with `FunctionCall(execute())`
2. and the `fits_on_single` fires here:
https://github.com/foundry-rs/foundry/blob/2e25e08d64169df2dc541b4d0a15b5790d068a96/fmt/src/formatter.rs#L1153

but couldn't make any additional progress because not familiar with the assumptions here, could please take another look at the `visit_block` function?

**Comment by DaniPopes** [comment]
Isn't (1) the intended behaviour? `visit_block` gets called with (true, true) so the single stmt gets inlined because it fits

**Comment by mattsse** [comment]
> Isn't (1) the intended behaviour? 

yes, was just pointing out the callgraph

> so the single stmt gets inlined because it fits

yes, but I didn't before, so something changed here

**Comment by DaniPopes** [comment]
The old manual implementation of `.loc()` for `FunctionDefinition` includes the block, while `solang-parser`'s actual parser excludes it. This breaks an internal assumption that there is only whitespace between `item[n].loc().end()` and `item[n + 1].loc().start()`

In `Formatter::write_lined_visitable`, `last_byte_written` gets assigned to `item.loc().end()`, and `unwritten_whitespace` returns the wrong value which includes the block:

```log
TRACE SU:function: forge_fmt::formatter: BEFORE: "<OMITTED>\n\n// forgefmt: disable-start\n\n"
TRACE SU:function: forge_fmt::formatter: AFTER:  "<OMITTED>\n\n// forgefmt: disable-start\n\nfunction test1() {}"
TRACE SU: forge_fmt::formatter: Unwritten whitespace: "{}\n\n"
TRACE SU:function: forge_fmt::formatter: BEFORE: "<OMITTED>\n\n// forgefmt: disable-start\n\nfunction test1() {}{}\n\n"
```

Should we revert to the old manual trait with just the special case for `FunctionDefinition`?

**Comment by mattsse** [comment]
> Should we revert to the old manual trait with just the special case for FunctionDefinition?

this sounds reasonable, would give us a bit more flexibility

**Comment by DaniPopes** [comment]
Done, this fixes point number 2.

Number 1 is fixed with:

```diff
diff --git a/fmt/testdata/IfStatement/fmt.sol b/fmt/testdata/IfStatement/fmt.sol
index 4c430108..cb2f8874 100644
--- a/fmt/testdata/IfStatement/fmt.sol
+++ b/fmt/testdata/IfStatement/fmt.sol
@@ -104,9 +104,7 @@ contract IfStatement {
             execute();
         }
 
-        if (condition && ((condition || anotherLongCondition))) {
-            execute();
-        }
+        if (condition && ((condition || anotherLongCondition))) execute();
 
         // if statement
         if (condition) execute();
diff --git a/fmt/testdata/IfStatement2/fmt.sol b/fmt/testdata/IfStatement2/fmt.sol
index f9c27b1f..10ae4360 100644
--- a/fmt/testdata/IfStatement2/fmt.sol
+++ b/fmt/testdata/IfStatement2/fmt.sol
@@ -2,8 +2,6 @@ contract IfStatement {
     function test() external {
         bool anotherLongCondition;
 
-        if (condition && ((condition || anotherLongCondition))) {
-            execute();
-        }
+        if (condition && ((condition || anotherLongCondition))) execute();
     }
 }
```

But I guess the question here is still why did this change.

File: fmt/src/formatter.rs [hunk 1]
```diff
@@ -152,10 +151,18 @@ impl<'a, W: Write> Formatter<'a, W> {
...
+    }
+
+    /// Casts the current `W` writer as a `String` reference. Should only be used for debugging.
+    #[allow(dead_code)]
+    unsafe fn buf_contents(&self) -> &String {
+        match &self.temp_bufs[..] {
+            [] => unsafe { *(&self.buf.w as *const W as *const &mut String) },
+            [.., buf] => &buf.w,
         }
     }
```
Comment (DaniPopes) [inline fmt/src/formatter.rs:167]: Used inline with `trace!("{}", self.buf_contents())`

This avoids adding `fmt::Debug` bounds everywhere, makes debugging faster. LMK if this should be removed.

File: fmt/src/formatter.rs [hunk 2]
```diff
@@ -3487,207 +3538,8 @@ impl<'a, W: Write> Visitor for Formatter<'a, W> {
...
 
+    #[instrument(name = "parser_error", skip_all)]
     fn visit_parser_error(&mut self, loc: Loc) -> Result<()> {
         Err(FormatterError::InvalidParsedItem(loc))
     }
 }
-
-#[cfg(test)]
-mod tests {
```
Comment (DaniPopes) [inline fmt/src/formatter.rs:3496]: moved to `../tests/it/formatter.rs`

File: fmt/src/solang_ext/loc.rs
Comment (DaniPopes) [inline fmt/src/solang_ext/loc.rs:1]: diff in: 815e1d8

**Comment by DaniPopes** [comment]
Some failing anvil test unrelated

**Comment by DaniPopes** [comment]
yep honestly i don't know why, i think it's fine just keeping it

**Comment by DaniPopes** [comment]
can you apply patch i posted above? i can't for the next few hours

**Pull request merged**